octave/fsk_cml_sam.m

```diff
@@ -0,0 +1,275 @@
+%    3   Some simple HD conversions
+%    4   CML approach using symbol likelihoods, then converted to bit LLRs
+%
+% Results from this sim are stored in "res" -- use fsk_llr_plot to see BER figs
+% Use "plt" to see some useful plots (for selected Ltypes) :
+%      1 is SD pdf histograms; 2 is Rx PSD;  3 is bit LLR histograms
+%
+% Adjust Evec and Nbits as required before running. 
+
+
+ldpc;
+
+% define Rician pdf
+function y = rice(x,v,s)
+  s2 = s*s; 
+  y = (x / s2) .* exp(-0.5 * (x.^2 + v.^2)/s2) .* besseli(0, x*v/s2);
+endfunction
+
+function plot_pdf(v,s)
+  x=(0:0.1:2*v); 
+  y= rice(x, v, s); 
+  figure(201);   hold on 
+  plot(x,y,'g');
+  %title('Rician pdf: signal carrier')
+  y= rice(x, 0, s); 
+  plot(x,y,'b');
+  title('Rician pdf: signal and noise-only carriers')
+  pause(0.01); 
+endfunction 
+  
+
+% single Eb/No point simulation    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+function [raw_ber rx_filt rx_bits tx_symbols demapper sig_est SNRest v_est] = ...
+	 	  run_single(tx_bits, M, EcNodB,  plt)
+% Ec/N0 is per channel bit
+bps = log2(M);  % bits per symbol
+Ts = 16;        % length of each symbol in samples
+
+if length(tx_bits)==1
+    Nbits = tx_bits;
+    tx_bits = randi(2,1,Nbits)-1;    % make random bits
+end
+
+Nbits = length(tx_bits);
+Nsymbols = Nbits/log2(M);
+tx_symbols = zeros(1,Nsymbols);
+
+mapper = bps:-1:1;
+% look up table demapper from symbols to bits (hard decision)
+demapper=zeros(M,bps);
+for m=1:M
+    for b=1:bps
+        if  bitand(m-1,b) demapper(m,bps-b+1) = 1; end
+    end
+end
+
+% continuous phase mFSK modulator
+
+w(1:M) = 2*pi*(1:M)/Ts;
+tx_phase = 0;
+tx = zeros(1,Ts*Nsymbols);
+
+for s=1:Nsymbols
+    bits_for_this_symbol = tx_bits(bps*(s-1)+1:bps*s);
+    symbol_index = bits_for_this_symbol * mapper' + 1;
+    tx_symbols(s) = symbol_index;
+    assert(demapper(symbol_index,:) == bits_for_this_symbol);
+    for k=1:Ts
+        tx_phase = tx_phase + w(symbol_index);
+        tx((s-1)*Ts+k) = exp(j*tx_phase);
+    end
+end
+
+% AWGN channel noise
+
+
+EsNodB = EcNodB + 10*log10(bps);
+EsNo = 10^(EsNodB/10);
+variance = Ts/EsNo;
+noise = sqrt(variance/2)*(randn(1,Nsymbols*Ts) + j*randn(1,Nsymbols*Ts));
+rx = tx + noise;
+
+if plt==2,    % check the Spectrum
+    [psd,Fpsd] =pwelch(rx,128,0.5,128,Ts);
+    figure(110); plot(Fpsd,10*log10(psd));
+    title('Rx Signal:  PSD ');
+    xlabel('Freq/Rs');
+    %figure(111);plot(unwrap(arg(tx)));
+    pause(0.01);
+end
+
+
+% integrate and dump demodulator
+
+rx_bits = zeros(1,Nbits);
+rx_filt = zeros(Nsymbols,M);
+rx_pows = zeros(1,M);
+rx_nse_pow  = 0.0; rx_sig_pow =0.0;
+for s=1:Nsymbols
+    arx_symb = rx((s-1)*Ts + (1:Ts));
+    for m=1:M
+        r= sum(exp(-j*w(m)*(1:Ts)) .* arx_symb);
+        rx_pows(m)= r * conj(r);
+        rx_filt(s,m) = abs(r);
+    end
+    [tmp symbol_index] = max(rx_filt(s,:));
+    rx_sig_pow = rx_sig_pow + rx_pows(symbol_index);
+    rx_pows(symbol_index)=[];
+    rx_nse_pow = rx_nse_pow + sum(rx_pows)/(M-1);
```

Comment:
So the noise est is the sum of all the other filter energies apart from the chosen (max) symbol?

Reply:
Yes